src/pudl/io_managers.py

```diff
@@ -0,0 +1,592 @@
+
+        # Create the database and schemas
+        if not db_path.exists():
+            db_path.touch()
+            self.md.create_all(engine)
+
+        return engine
+
+    def _get_sqlalchemy_table(self, table_name: str) -> sa.Table:
+        """Get SQL Alchemy Table object from metadata given a table_name.
+
+        Args:
+            table_name: The name of the table to look up.
+
+        Returns:
+            table: Corresponding SQL Alchemy Table in SQLiteIOManager metadata.
+
+        Raises:
+            RuntimeError: if table_name does not exist in the SQLiteIOManager metadata.
+        """
+        sa_table = self.md.tables.get(table_name, None)
+        if sa_table is None:
+            # TODO (bendnorman): Logging a warning for now so the analysis example can run but we could raise an error.
+            # raise RuntimeError(
+            #     f"{sa_table} not found in database metadata. Either add the table to the metadata or use a different IO Manager."
+            # )
+            logger.warning(
+                f"{table_name} not found in database metadata. Dtypes of returned DataFrame might be incorrect."
+            )
+        return sa_table
+
+    def _get_fk_list(self, table: str) -> pd.DataFrame:
+        """Retrieve a dataframe of foreign keys for a table.
+
+        Description from the SQLite Docs:
+        'This pragma returns one row for each foreign key constraint
+        created by a REFERENCES clause in the CREATE TABLE statement of table "table-name".'
+
+        The PRAGMA returns one row for each field in a foreign key constraint.
+        This method collapses foreign keys with multiple fields into one record
+        for readability.
+        """
+        with self.engine.connect() as con:
+            table_fks = pd.read_sql_query(f"PRAGMA foreign_key_list({table});", con)
+
+        # Foreign keys with multiple fields are reported in separate records.
+        # Combine the multiple fields into one string for readability.
+        # Drop duplicates so we have one FK for each table and foreign key id
+        table_fks["fk"] = table_fks.groupby("table")["to"].transform(
+            lambda field: "(" + ", ".join(field) + ")"
+        )
+        table_fks = table_fks[["id", "table", "fk"]].drop_duplicates()
+
+        # Rename the fields so we can easily merge with the foreign key errors.
+        table_fks = table_fks.rename(columns={"id": "fkid", "table": "parent"})
+        table_fks["table"] = table
+        return table_fks
+
+    def check_foreign_keys(self) -> None:
+        """Check foreign key relationships in the database.
+
+        The order assets are loaded into the database will not satisfy foreign key
+        constraints so we can't enable foreign key constraints. However, we can
+        check for foreign key failures once all of the data has been loaded into
+        the database using the `foreign_key_check` and `foreign_key_list` PRAGMAs.
```

Comment:
Ah, clever! Nice workaround.

Reply:
Thanks! Checking referential integrity in the test suite instead of at load time seems like an acceptable pattern for data warehouses. For example, [dbt allows you to create referential integrity tests](https://docs.getdbt.com/docs/build/tests) that are run once the data is loaded in.

I imagined running adding this FK test to the CI so we are regularly checking we're not breaking any of the data relationships.

The `foreign_key_check` pragma is super useful for our situation but I don't think it is common in other databases. If we ever decide to move away from SQLite we could recreate the referential integrity test by adding FK constraints to the tables after the data is loaded.